core/src/main/java/org/apache/iceberg/avro/AvroWithPartnerByStructureVisitor.java

```diff
@@ -93,19 +94,41 @@ private static <P, T> T visitRecord(
   private static <P, T> T visitUnion(
       P type, Schema union, AvroWithPartnerByStructureVisitor<P, T> visitor) {
     List<Schema> types = union.getTypes();
-    Preconditions.checkArgument(
-        AvroSchemaUtil.isOptionSchema(union), "Cannot visit non-option union: %s", union);
     List<T> options = Lists.newArrayListWithExpectedSize(types.size());
-    for (Schema branch : types) {
-      if (branch.getType() == Schema.Type.NULL) {
-        options.add(visit(visitor.nullType(), branch, visitor));
-      } else {
-        options.add(visit(type, branch, visitor));
+    if (AvroSchemaUtil.isOptionSchema(union)) {
+      for (Schema branch : types) {
+        if (branch.getType() == Schema.Type.NULL) {
+          options.add(visit(visitor.nullType(), branch, visitor));
+        } else {
+          options.add(visit(type, branch, visitor));
+        }
+      }
+    } else {
+      Types.StructType struct = (Types.StructType) type;
+      int index = 0;
+      for (Schema branch : types) {
+        if (branch.getType() != Schema.Type.NULL) {
+          options.add(visit(fieldTypeByName(type, "field" + index, visitor), branch, visitor));
```

Comment:
This visitor is by structure and not by name. That should be sufficient for the purpose of this PR, because we know the structure of the schemas match (one was just converted from the other).

Can you remove the by-name code here? It doesn't make sense for this visitor and I don't think we need it for the PR.

Reply:
Reason for going by the name was that the name also reflects a position (which in some cases is more authoritative), but it could work either way. Removed the name.